src/optimization/Combiner.cpp

```diff
@@ -1121,6 +1125,518 @@ InstructionWalker optimizations::combineArithmeticOperations(
+        }
+    }
+
+    // SubExpression result(INT_ZERO);
+    // for(auto& p : expanded)
+    // {
+    //     result = SubExpression(std::make_shared<Expression>(p.first ? OP_ADD : OP_SUB, result, p.second));
+    // }
+    //
+    // return result;
+}
+
+SubExpression replaceLocalToExpr(const SubExpression& subExpr, const Value& local, SubExpression newExpr)
+{
+    if(auto expr = subExpr.checkExpression())
+    {
+        return SubExpression(std::make_shared<Expression>(expr->code,
+                    replaceLocalToExpr(expr->arg0, local, newExpr),
+                    replaceLocalToExpr(expr->arg1, local, newExpr)));
+
+    }
+    else if(auto replacee = subExpr.checkLocal())
+    {
+        if (auto replacer = local.checkLocal()) {
+            if (*replacee == *replacer) {
+                return newExpr;
+            }
+        }
+    }
+
+    return subExpr;
+}
+
+void optimizations::combineDMALoads(const Module& module, Method& method, const Configuration& config)
+{
+    using namespace std;
+    using namespace VariantNamespace;
+
+    const std::regex vloadReg("vload(2|3|4|8|16)");
+
+    for(auto& bb : method)
+    {
+        //             loadInstrs,                        offsetValues,  addrValue
+        map<int, tuple<vector<intermediate::MethodCall*>, vector<Value>, Optional<Value>>> vloads;
+
+        for(auto& it : bb)
+        {
+            // Find all vloadn calls
+            if(auto call = dynamic_cast<intermediate::MethodCall*>(it.get()))
+            {
+                auto name = vc4c::spirv::demangleFunctionName(call->methodName);
+
+                std::smatch m;
+                if(std::regex_search(name, m, vloadReg))
+                {
+                    int n = std::stoi(m.str(1));
+
+                    // TODO: Check whether all second argument values are equal.
+
+                    auto& vload = vloads[n];
+                    auto& loadInstrs = get<0>(vload);
+                    auto& offsetValues = get<1>(vload);
+                    auto& addrValue = get<2>(vload);
+
+                    if(!addrValue.has_value())
+                    {
+                        addrValue = call->getArgument(1);
+                    }
+                    else if(addrValue != call->getArgument(1))
+                    {
+                        continue;
+                    }
+
+                    offsetValues.push_back(call->assertArgument(0));
+                    loadInstrs.push_back(call);
+                }
+            }
+        }
+
+        for(auto& p : vloads)
+        {
+            auto vectorLength = p.first;
+            auto& vload = p.second;
+            auto& loadInstrs = get<0>(vload);
+            auto& offsetValues = get<1>(vload);
+            auto& addrValue = get<2>(vload);
+
+            if(offsetValues.size() <= 1)
+                continue;
+
+            for(auto& inst : loadInstrs)
+            {
+                logging::debug() << inst->to_string() << logging::endl;
```

Comment:
This needs some text to tell the reader of the log, what this value means, e.g.

    logging::debug() << "Combining vload instruction: " << inst->to_string() << logging::endl